File: jsengine-android/jsengine/src/main/java/com/cliqz/jsengine/Adblocker.java
package com.cliqz.jsengine;

import android.util.Log;

import com.cliqz.jsengine.v8.V8Engine;
import com.eclipsesource.v8.V8;
import com.eclipsesource.v8.V8Object;
import com.eclipsesource.v8.debug.mirror.StringMirror;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeoutException;

/**
 * Created by sammacbeth on 27/10/2016.
 */

public class Adblocker {

    private final static String TAG = Adblocker.class.getSimpleName();

    private final Engine engine;

    private final static String MODULE_NAME = "adblocker";
    private final static String ENABLE_PREF = "cliqz-adb";
    private final static String ABTEST_PREF = "cliqz-adb-abtest";

    public Adblocker(Engine engine) {
        this.engine = engine;
    }

    /**
     * Get default prefences to load the adblock module in a disabled state
     * @return
     */
    public static Map<String, Object> getDefaultPrefs() {
        return getDefaultPrefs(false);
    }

    /**
     * Get the default preferences to load the adblock module in the state given by enabled
     * @param enabled
     * @return
     */
    public static Map<String, Object> getDefaultPrefs(final boolean enabled) {
        final Map<String, Object> prefs = new HashMap<>();
        prefs.put(ABTEST_PREF, true);
        prefs.put(ENABLE_PREF, 1);
        prefs.put("modules."+ MODULE_NAME + ".enabled", enabled);
        return prefs;
    }

    public void setEnabled(final boolean enabled) throws ExecutionException {
        engine.setPref(ABTEST_PREF, true);
        engine.setPref(ENABLE_PREF, enabled ? 1 : 0);
        engine.setPref("modules."+ MODULE_NAME + ".enabled", enabled);
    }

    public JSONObject getAdBlockingInfo(final String url) {
        try {
            final Object stats = engine.system.callFunctionOnModuleAttribute(MODULE_NAME + "/adblocker", new String[]{"default", "adbStats"}, "report", url);
            return new JSONObject(stats.toString());
        } catch(ExecutionException | JSONException e) {
//            Log.e(TAG, "getAdBlockingInfo", e);
        }
        return new JSONObject();
    }

    public boolean isBlacklisted(final String url) {
        try {
            final Object blacklisted = engine.system.callFunctionOnModuleAttribute(50, MODULE_NAME +"/adblocker", new String[]{"default", "adBlocker"}, "isDomainInBlacklist", url);
            return blacklisted.equals(Boolean.TRUE);
        } catch(ExecutionException e) {
//            Log.e(TAG, "isBlacklisted", e);
        } catch(TimeoutException e) {
            return false;
        }
        return false;
    }

    public void toggleUrl(final String url, final boolean domain) {
        try {
            engine.system.callFunctionOnModuleAttribute(MODULE_NAME +"/adblocker", new String[]{"default", "adBlocker"}, "toggleUrl", url, domain);
        } catch(ExecutionException e) {
//            Log.e(TAG, "toggleUrl", e);
        }
    }

}

File: jsengine-ios/jsengine/Adblocker.swift
//
//  Adblocker.swift
//  jsengine
//
//  Created by Ghadir Eraisha on 1/16/17.
//  Copyright © 2017 Cliqz GmbH. All rights reserved.
//

import Foundation

public class Adblocker {
    
    //MARK: - Constants
    private static let moduleName = "adblocker"
    private static let enablePref = "cliqz-adb"
    private static let abTestPref = "cliqz-adb-abtest"
    
    //MARK: - Instant variables
    var engine: Engine
    
    //MARK: - Init
    public init(engine: Engine) {
        self.engine = engine
    }
    
    //MARK: - Public APIs
    public class func getDefaultPrefs(enabled: Bool? = true) -> [String: AnyObject] {
        var prefs: [String:AnyObject] = [String:AnyObject]()
        prefs[Adblocker.enablePref] = 1
        prefs[Adblocker.abTestPref] = true
        prefs["modules." + Adblocker.moduleName + ".enabled"] = enabled
        
        return prefs
    }
    
    public func setEnabled(enabled: Bool) {
        dispatch_async(self.engine.dispatchQueue) {
            self.engine.setPref(Adblocker.abTestPref, prefValue: true)
            self.engine.setPref(Adblocker.enablePref, prefValue: enabled ? 1 : 0)
            self.engine.setPref("modules." + Adblocker.moduleName + ".enabled", prefValue: enabled)
        }
    }
    
    public func getAdsCounter(url: String) -> Int? {
        guard self.engine.isRunning() else {
            return nil
        }
        
        let blockingInfo = getAdBlockingInfo(url)
        return blockingInfo!["totalCount"] as? Int
    }
    
    public func getAdBlockingInfo(url: String) -> [NSObject : AnyObject]? {
        guard self.engine.isRunning() else {
            return nil
        }
        
        do {
            let stats = try engine.systemLoader?.callFunctionOnModuleAttribute(Adblocker.moduleName + "/adblocker", attribute: ["default", "adbStats"], functionName: "report", arguments: [url])
            return stats?.toDictionary()
        } catch let error as NSError {
            DebugLogger.log("<< Error in Adblocker.getAdBlockingInfo: \(error)")
        }
        return nil
    }
    
    public func isBlacklisted(url: String) -> Bool? {
        guard self.engine.isRunning() else {
            return nil
        }
        
        do {
            let blacklisted = try engine.systemLoader?.callFunctionOnModuleAttribute(Adblocker.moduleName + "/adblocker", attribute: ["default", "adBlocker"], functionName: "isDomainInBlacklist", arguments: [url])
            if let blacklisted =  blacklisted {
                return blacklisted.toBool()
            }
        } catch let error as NSError {
            DebugLogger.log("<< Error in Adblocker.isBlacklisted: \(error)")
        }
        return false
    }
    
    public func toggleUrl(url: String, domain: Bool) {
        do {
            try self.engine.systemLoader?.callFunctionOnModuleAttribute(Adblocker.moduleName + "/adblocker", attribute: ["default", "adBlocker"], functionName: "toggleUrl", arguments: [url, domain])
        } catch let error as NSError {
            DebugLogger.log("<< Error in Adblocker.toggleUrl: \(error)")
        }
    }
}
